src/service/VotingService.ts

```diff
@@ -14,108 +14,107 @@
 
-type VotingParams = ConfigParams;
-
-export interface VotingMetadata {
-    readonly votingKeyStartEpoch: number;
-    readonly votingKeyEndEpoch: number;
-    readonly votingPublicKey: string;
-    readonly version: number;
-}
+type VotingParams = { target: string; user: string };
 
 const logger: Logger = LoggerFactory.getLogger(LogType.System);
 
 export class VotingService {
-    private static readonly METADATA_VERSION = 1;
-
     constructor(protected readonly params: VotingParams) {}
 
-    public async run(presetData: ConfigPreset, nodeAccount: NodeAccount, nodePreset: NodePreset | undefined): Promise<void> {
+    public async run(
+        presetData: ConfigPreset,
+        nodeAccount: NodeAccount,
+        nodePreset: NodePreset,
+        currentNetworkEpoch: number | undefined,
+        updateVotingKey: boolean | undefined,
+    ): Promise<boolean> {
         const symbolServerImage = presetData.symbolServerImage;
-
-        if (nodePreset?.voting && nodeAccount.voting) {
-            const privateKeyTreeFileName = 'private_key_tree1.dat';
+        const networkEpoch = currentNetworkEpoch || presetData.lastKnownNetworkEpoch || 1;
+        const update = updateVotingKey === undefined ? presetData.autoUpdateVotingKeys : updateVotingKey;
+        if (nodePreset?.voting) {
             const target = this.params.target;
             const votingKeysFolder = join(
                 BootstrapUtils.getTargetNodesFolder(target, true, nodeAccount.name),
                 presetData.votingKeysDirectory,
             );
-            const metadataFile = join(votingKeysFolder, 'metadata.yml');
-            if (!(await this.shouldGenerateVoting(presetData, metadataFile, nodeAccount.voting))) {
-                logger.info(`Voting File for node ${nodePreset.name} has been previously generated. Reusing...`);
-                return;
-            }
-            const votingPrivateKey = nodeAccount?.voting.privateKey;
-
-            if (!votingPrivateKey) {
-                throw new Error(
-                    'Voting key should have been previously generated!!! You need to reset your target folder. Please run --reset using your original custom preset.',
-                );
-            }
-
-            const cmd = [
-                `${presetData.catapultAppFolder}/bin/catapult.tools.votingkey`,
-                `--secret=${votingPrivateKey}`,
-                `--startEpoch=${presetData.votingKeyStartEpoch}`,
-                `--endEpoch=${presetData.votingKeyEndEpoch}`,
-                `--output=/votingKeys/${privateKeyTreeFileName}`,
-            ];
-            await BootstrapUtils.deleteFolder(votingKeysFolder);
             await BootstrapUtils.mkdir(votingKeysFolder);
-            const binds = [`${votingKeysFolder}:/votingKeys:rw`];
+            const votingUtils = new VotingUtils();
+            await BootstrapUtils.deleteFile(join(votingKeysFolder, 'metadata.yml'));
+            let newFileCreated = false;
+            while (true) {
```

Comment:
Not sure why ``while(true)``?  maybe I am missing something.
Are multiple files getting created?  I would assume when the loop is finish either it creates a voting key file or not.

Reply:
if you have a short votingKeyDesiredLifetime and a large votingKeyDesiredFutureLifetime in your custom preset, the service generates multiple files. Most people would not change this, but you have the option of having smaller files at once if you like

Reply:
ok.  I would suggest not generating keys that are not been used.

Reply:
Fair enough. I have removed the while and only one file can be created at a given time. Also, votingKeyDesiredFutureLifetime cannot be greater than votingKeyDesiredLifetime.